lib/tree.rb

```diff
@@ -16,42 +16,136 @@ def initialize
+    # LOOP WAY
+    # if @root.nil? 
+    #   @root = TreeNode.new(key, value)
+    # else 
+    #   current = @root 
+    #   while true 
+    #     if key < current.key 
+    #       if !current.left.nil? 
+    #         current = current.left
+    #       else 
+    #         current.left = TreeNode.new(key, value)
+    #       end 
+    #     else 
+    #       if !current.right.nil? 
+    #         current = current.right
+    #       else 
+    #         current.right = TreeNode.new(key, value)
+    #       end 
+    #     end 
+    #   end 
+    # end 
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # helper method for add() recursive way 
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil? 
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else 
+      current_node.right = add_helper(current_node.right, key, value)
+    end 
+
+    return current_node
+  end 
+
+
+  # Time Complexity: o(log n)
+  # Space Complexity: 0(n)
   def find(key)
-    raise NotImplementedError
+    find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+
+    if current_node.key == key
+      return current_node.value 
+    elsif key < current_node.key
+      return find_helper(current_node.left, key)
+    else 
+      return find_helper(current_node.right, key)
+    end 
+    return current_node
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current_node, list)
+    return list if current_node.nil? 
+
+    inorder_helper(current_node.left, list)
+    list << {key: current_node.key, value: current_node.value}
+    inorder_helper(current_node.right, list)
+
+    return list 
+  end 
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    return list if current_node.nil? 
+
+    list << {key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+
+    return list 
+  end 
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current_node, list)
+    return list if current_node.nil? 
+
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << {key: current_node.key, value: current_node.value}
+
+    return list 
+  end 
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
Your time complexity is right, and the space complexity is right if the tree is unbalanced.